File: legacy_backend/aggregators/mactechnews.py
"""Aggregator for MacTechNews RSS feeds."""

from bs4 import BeautifulSoup
from pydantic import BaseModel, Field

from .base import BaseAggregator, RawArticle


class MacTechNewsAggregatorConfig(BaseModel):
    id: str
    type: str = Field(pattern="^(managed|custom|social)$")
    name: str = Field(min_length=1)
    url: str = ""
    description: str = Field(min_length=1)
    wait_for_selector: str | None = None
    selectors_to_remove: list[str] = Field(default_factory=list)
    options: dict = Field(default_factory=dict)


class MacTechNewsAggregator(BaseAggregator):
    """Aggregator for MacTechNews.de (German Apple news)."""

    id = "mactechnews"
    type = "managed"
    name = "MacTechNews"
    url = "https://www.mactechnews.de/Rss/News.x"
    description = "Specialized aggregator for MacTechNews.de (German Apple news). Extracts article content from MtnArticle elements, removes mobile headers and sidebars."
    wait_for_selector = ".MtnArticle"
    selectors_to_remove = [
        ".NewsPictureMobile",
        "header",
        "aside",
        "script",
        "style",
        "iframe",
        "noscript",
        "svg",
    ]

    def __init__(self):
        super().__init__()
        MacTechNewsAggregatorConfig(
            id=self.id,
            type=self.type,
            name=self.name,
            url=self.url,
            description=self.description,
            wait_for_selector=self.wait_for_selector,
            selectors_to_remove=self.selectors_to_remove,
            options=self.options,
        )

    def extract_content(self, article: RawArticle) -> None:
        """Extract content from .MtnArticle element."""
        try:
            soup = BeautifulSoup(article.html, "html.parser")
            article_content = soup.select_one(".MtnArticle")
            if not article_content:
                self.logger.warning(
                    f"Could not find .MtnArticle content in {article.url}"
                )
                return
            for tag in article_content.find_all(["p", "div", "span"]):
                if not tag.get_text(strip=True) and not tag.find("img"):
                    tag.decompose()
            article.html = str(article_content)
        except Exception as e:
            self.logger.error(
                f"Extraction failed for {article.url}: {e}", exc_info=True
            )


def aggregate(feed, force_refresh=False, options=None):
    aggregator = MacTechNewsAggregator()
    return aggregator.aggregate(feed, force_refresh, options or {})

File: src/server/aggregators/mactechnews.ts
/**
 * MacTechNews aggregator.
 *
 * Specialized aggregator for MacTechNews.de (German Apple news).
 * Extracts article content from .MtnArticle elements, removes mobile headers and sidebars.
 */

import { BaseAggregator } from "./base/aggregator";
import type { RawArticle } from "./base/types";
import { fetchFeed } from "./base/fetch";
import Parser from "rss-parser";

export class MacTechNewsAggregator extends BaseAggregator {
  override readonly id: string = "mactechnews";
  override readonly type: "managed" | "custom" | "social" = "managed";
  override readonly name: string = "MacTechNews";
  override readonly url: string = "https://www.mactechnews.de/Rss/News.x";
  override readonly description: string =
    "MacTechNews.de - German technology news website focused on Apple products and ecosystem.";

  override readonly selectorsToRemove: string[] = [
    ".NewsPictureMobile",
    "header",
    "aside",
    "script",
    "style",
    "iframe",
    "noscript",
    "svg",
  ];

  override readonly waitForSelector: string = ".MtnArticle";

  /**
   * Fetch RSS feed data.
   */
  protected override async fetchSourceData(
    limit?: number,
  ): Promise<Parser.Output<any>> {
    const startTime = Date.now();
    this.logger.info(
      {
        step: "fetchSourceData",
        subStep: "start",
        aggregator: this.id,
        feedId: this.feed?.id,
        limit,
      },
      "Fetching RSS feed",
    );

    if (!this.feed) {
      throw new Error("Feed not initialized");
    }

    const feedUrl = this.feed.identifier || this.url;
    const feed = await fetchFeed(feedUrl);

    const elapsed = Date.now() - startTime;
    this.logger.info(
      {
        step: "fetchSourceData",
        subStep: "complete",
        aggregator: this.id,
        feedId: this.feed?.id,
        itemCount: feed.items?.length || 0,
        elapsed,
      },
      "RSS feed fetched",
    );

    return feed;
  }

  /**
   * Parse RSS feed items to RawArticle[].
   */
  protected override async parseToRawArticles(
    sourceData: unknown,
  ): Promise<RawArticle[]> {
    const startTime = Date.now();
    this.logger.info(
      {
        step: "parseToRawArticles",
        subStep: "start",
        aggregator: this.id,
        feedId: this.feed?.id,
      },
      "Parsing RSS feed items",
    );

    const feed = sourceData as Parser.Output<any>;
    const items = feed.items || [];

    const articles: RawArticle[] = items.map((item) => ({
      title: item.title || "",
      url: item.link || "",
      published: item.pubDate ? new Date(item.pubDate) : new Date(),
      summary: item.contentSnippet || item.content || "",
      author: item.creator || item.author || undefined,
    }));

    const elapsed = Date.now() - startTime;
    this.logger.info(
      {
        step: "parseToRawArticles",
        subStep: "complete",
        aggregator: this.id,
        feedId: this.feed?.id,
        articleCount: articles.length,
        elapsed,
      },
      "Parsed RSS feed items",
    );

    return articles;
  }

  /**
   * Extract content from .MtnArticle element.
   */
  protected override async extractContent(
    html: string,
    article: RawArticle,
  ): Promise<string> {
    const startTime = Date.now();
    this.logger.debug(
      {
        step: "extractContent",
        subStep: "extractMtnArticle",
        aggregator: this.id,
        feedId: this.feed?.id,
        url: article.url,
      },
      "Extracting content from .MtnArticle element",
    );

    const { extractContent } = await import("./base/extract");
    const extracted = extractContent(html, {
      contentSelector: ".MtnArticle",
      selectorsToRemove: this.selectorsToRemove,
    });

    // Use base removeElementsBySelectors for additional cleanup
    const result = await super.removeElementsBySelectors(extracted, article);

    const elapsed = Date.now() - startTime;
    this.logger.debug(
      {
        step: "extractContent",
        subStep: "extractMtnArticle",
        aggregator: this.id,
        feedId: this.feed?.id,
        url: article.url,
        elapsed,
      },
      "Content extracted from .MtnArticle",
    );

    return result;
  }
}
